File: frontend/src/utils/regexCompat.ts
const SAFE_REGEX_LITERAL_FLAGS = new Set(['i', 'm', 's', 'u'])

export interface ParsedRegexInput {
  pattern: string
  flags: string
  normalizedFromLiteral: boolean
}

export function parseRegexInput(raw: string): ParsedRegexInput {
  const text = (raw || '').trim()
  if (text.length < 2 || text[0] !== '/') {
    return { pattern: text, flags: '', normalizedFromLiteral: false }
  }

  let inClass = false
  for (let idx = 1; idx < text.length; idx += 1) {
    const ch = text[idx]
    if (ch === '\\') {
      idx += 1
      continue
    }
    if (ch === '[' && !inClass) {
      inClass = true
      continue
    }
    if (ch === ']' && inClass) {
      inClass = false
      continue
    }
    if (ch === '/' && !inClass) {
      const pattern = text.slice(1, idx)
      const suffix = text.slice(idx + 1)
      if (suffix && !/^[a-z]+$/i.test(suffix)) break

      let flags = ''
      const seen = new Set<string>()
      for (const rawFlag of suffix.toLowerCase()) {
        if (seen.has(rawFlag) || !SAFE_REGEX_LITERAL_FLAGS.has(rawFlag)) {
          return { pattern: text, flags: '', normalizedFromLiteral: false }
        }
        seen.add(rawFlag)
        flags += rawFlag
      }
      return { pattern, flags, normalizedFromLiteral: true }
    }
  }

  return { pattern: text, flags: '', normalizedFromLiteral: false }
}

export function buildRegexFromInput(raw: string): RegExp {
  const parsed = parseRegexInput(raw)
  return new RegExp(parsed.pattern, parsed.flags)
}

File: backend/app/regex_compat.py
"""
世界书正则兼容辅助：

- 保持存储的 regex 原样不变
- 仅在输入整体形如 /.../flags 且 flags 属于安全白名单时，按 JS 字面量拆壳
- Python 侧只映射可安全落地的 flags，避免静默改变语义
"""

from __future__ import annotations

import re

_SAFE_LITERAL_FLAG_MAP: dict[str, int] = {
    "i": re.IGNORECASE,
    "m": re.MULTILINE,
    "s": re.DOTALL,
    "u": 0,
}


def split_regex_literal(raw: str) -> tuple[str, int] | None:
    text = (raw or "").strip()
    if len(text) < 2 or not text.startswith("/"):
        return None

    in_class = False
    idx = 1
    while idx < len(text):
        ch = text[idx]
        if ch == "\\":
            idx += 2
            continue
        if ch == "[" and not in_class:
            in_class = True
            idx += 1
            continue
        if ch == "]" and in_class:
            in_class = False
            idx += 1
            continue
        if ch == "/" and not in_class:
            body = text[1:idx]
            suffix = text[idx + 1 :]
            if suffix and not suffix.isalpha():
                return None
            flags = 0
            seen: set[str] = set()
            for raw_flag in suffix.lower():
                if raw_flag in seen:
                    return None
                mapped = _SAFE_LITERAL_FLAG_MAP.get(raw_flag)
                if mapped is None:
                    return None
                seen.add(raw_flag)
                flags |= mapped
            return body, flags
        idx += 1
    return None


def compile_user_regex(raw: str, base_flags: int = 0) -> re.Pattern[str]:
    parsed = split_regex_literal(raw)
    if parsed is None:
        return re.compile((raw or "").strip(), base_flags)
    pattern, literal_flags = parsed
    return re.compile(pattern, base_flags | literal_flags)
